projects/bp-gallery/src/components/views/picture/face-tagging/helpers/image-rect.ts

```diff
@@ -0,0 +1,140 @@
+import { useCallback, useEffect, useMemo, useState } from 'react';
+
+export const useImageRect = (img: HTMLImageElement | null) => {
+  const computeRect = useCallback((img: HTMLImageElement) => {
+    const { x, y, width, height } = img.getBoundingClientRect();
+    const { naturalHeight, naturalWidth } = img;
+    const naturalRatio = naturalWidth / naturalHeight;
+    const containerRatio = width / height;
+    const realWidth =
+      isNaN(naturalRatio) || naturalRatio > containerRatio ? width : height * naturalRatio;
+    const realHeight =
+      isNaN(naturalRatio) || naturalRatio < containerRatio ? height : width / naturalRatio;
+    const realX = x + (width - realWidth) / 2;
+    const realY = y + (height - realHeight) / 2;
+    return {
+      x: realX,
+      y: realY,
+      width: realWidth,
+      height: realHeight,
+    };
+  }, []);
+
+  const [rect, setRect] = useState(() => (img ? computeRect(img) : null));
+
+  const recompute = useCallback(
+    (img: HTMLImageElement) => {
+      setRect(computeRect(img));
+    },
+    [computeRect]
+  );
+
+  // The IntersectionObserver (see below) can only observe
```

Comment:
das sind mir irgendwie ein wenig viele kommentare. aber sicherlich war das auch nicht ganz einfach und trivial. vielleicht kann man dies in atomare Funktionen auslagern und dort kommentieren, sodass man wenn man den code nochmal anfassen muss, nicht unbedingt diese komplizierten Funktionen verändern muss

Reply:
Habs jetzt aufgeteilt. Die Kommentare habe ich aber da gelassen, wo ich glaube, dass sie am sinnvollsten sind, nämlich da, wo sie erklären, warum etwas so gemacht wurde wie es gemacht wurde.